File: src/python-app/src/presentation/formatters.py
"""Output formatting for terminal display."""

from typing import Final

from src.domain.entities import OperationResult, OperationStatus, ProcessingSummary


class ColorFormatter:
    """Formats messages with Rich terminal colors."""
    
    # Color templates - using ASCII-safe characters
    SUCCESS: Final[str] = "[green]+[/green]"
    ERROR: Final[str] = "[red]x[/red]"
    WARNING: Final[str] = "[yellow]![/yellow]"
    INFO: Final[str] = "[blue]i[/blue]"
    
    @classmethod
    def success(cls, message: str) -> str:
        """Format success message."""
        return f"[green]{message}[/green]"
    
    @classmethod
    def error(cls, message: str) -> str:
        """Format error message."""
        return f"[red]{message}[/red]"
    
    @classmethod
    def warning(cls, message: str) -> str:
        """Format warning message."""
        return f"[yellow]{message}[/yellow]"
    
    @classmethod
    def info(cls, message: str) -> str:
        """Format info message."""
        return f"[blue]{message}[/blue]"


class OperationResultFormatter:
    """Formats operation results for display."""
    
    @staticmethod
    def format(result: OperationResult) -> str:
        """
        Format an operation result.
        
        Args:
            result: Operation result to format
            
        Returns:
            Formatted string with color markup
        """
        record_value = result.record.value
        message = result.message
        
        if result.status == OperationStatus.SUCCESS:
            return f"{ColorFormatter.SUCCESS} {record_value}: {message}"

        if result.status == OperationStatus.UPDATED:
            return f"{ColorFormatter.SUCCESS} {record_value}: {message}"

        if result.status == OperationStatus.ALREADY_EXISTS:
            return f"{ColorFormatter.WARNING} {record_value}: {message}"

        if result.status == OperationStatus.SKIPPED:
            return f"{ColorFormatter.WARNING} {record_value}: Skipped (invalid format)"

        if result.status == OperationStatus.FAILED:
            return f"{ColorFormatter.ERROR} {record_value}: {message}"

        return f"{ColorFormatter.INFO} {record_value}: Unknown status"


class SummaryFormatter:
    """Formats processing summaries."""
    
    @staticmethod
    def format(summary: ProcessingSummary) -> str:
        """
        Format a processing summary.
        
        Args:
            summary: Processing summary to format
            
        Returns:
            Multi-line formatted summary
        """
        lines = [
            "\n" + "=" * 60,
            ColorFormatter.info("Processing Summary"),
            "=" * 60,
            f"Total records: {summary.total}",
            f"{ColorFormatter.SUCCESS} Created: {summary.successful}",
            f"{ColorFormatter.SUCCESS} Updated: {summary.updated}",
            f"{ColorFormatter.WARNING} Already existed: {summary.already_exists}",
            f"{ColorFormatter.ERROR} Failed: {summary.failed}",
            f"{ColorFormatter.WARNING} Skipped: {summary.skipped}",
            f"\nSuccess rate: {summary.success_rate:.1f}%",
            "=" * 60,
        ]
        
        return "\n".join(lines)


class GroupCreationFormatter:
    """Formats group creation results."""
    
    @staticmethod
    def format(group_name: str, created: bool) -> str:
        """
        Format group creation result.
        
        Args:
            group_name: Name of the group
            created: Whether the group was created or already existed
            
        Returns:
            Formatted string
        """
        if created:
            return f"{ColorFormatter.SUCCESS} Created group: {group_name}"
        
        return f"{ColorFormatter.WARNING} Group already exists: {group_name}"

File: src/rust-app/src/presentation/formatters.rs
//! Output formatting for terminal display. Ports `src/presentation/formatters.py`.
//!
//! Formatter functions return Rich-style markup strings (e.g. `[green]ok[/green]`)
//! exactly as the Python code does. `render_markup` converts that markup to ANSI
//! escape codes at the point of printing.

use crate::domain::entities::{OperationResult, OperationStatus, ProcessingSummary};

// Color marker templates (ASCII-safe), matching `ColorFormatter`.
pub const SUCCESS: &str = "[green]+[/green]";
pub const ERROR: &str = "[red]x[/red]";
pub const WARNING: &str = "[yellow]![/yellow]";
pub const INFO: &str = "[blue]i[/blue]";

/// Wrap a message in green markup.
pub fn success(message: &str) -> String {
    format!("[green]{message}[/green]")
}

/// Wrap a message in red markup.
pub fn error(message: &str) -> String {
    format!("[red]{message}[/red]")
}

/// Wrap a message in yellow markup.
pub fn warning(message: &str) -> String {
    format!("[yellow]{message}[/yellow]")
}

/// Wrap a message in blue markup.
pub fn info(message: &str) -> String {
    format!("[blue]{message}[/blue]")
}

/// Format a single operation result, matching `OperationResultFormatter`.
pub fn format_result(result: &OperationResult) -> String {
    let value = &result.record.value;
    let message = &result.message;
    match result.status {
        OperationStatus::Success | OperationStatus::Updated => {
            format!("{SUCCESS} {value}: {message}")
        }
        OperationStatus::AlreadyExists => format!("{WARNING} {value}: {message}"),
        OperationStatus::Skipped => format!("{WARNING} {value}: Skipped (invalid format)"),
        OperationStatus::Failed => format!("{ERROR} {value}: {message}"),
    }
}

/// Format a processing summary, matching `SummaryFormatter`.
pub fn format_summary(summary: &ProcessingSummary) -> String {
    let bar = "=".repeat(60);
    let lines = [
        format!("\n{bar}"),
        info("Processing Summary"),
        bar.clone(),
        format!("Total records: {}", summary.total),
        format!("{SUCCESS} Created: {}", summary.successful),
        format!("{SUCCESS} Updated: {}", summary.updated),
        format!("{WARNING} Already existed: {}", summary.already_exists),
        format!("{ERROR} Failed: {}", summary.failed),
        format!("{WARNING} Skipped: {}", summary.skipped),
        format!("\nSuccess rate: {:.1}%", summary.success_rate()),
        bar,
    ];
    lines.join("\n")
}

/// Format a group creation result, matching `GroupCreationFormatter`.
pub fn format_group(group_name: &str, created: bool) -> String {
    if created {
        format!("{SUCCESS} Created group: {group_name}")
    } else {
        format!("{WARNING} Group already exists: {group_name}")
    }
}

/// Convert Rich-style color markup into ANSI escape codes for terminal output.
pub fn render_markup(text: &str) -> String {
    text.replace("[green]", "\x1b[32m")
        .replace("[red]", "\x1b[31m")
        .replace("[yellow]", "\x1b[33m")
        .replace("[blue]", "\x1b[34m")
        .replace("[/green]", "\x1b[0m")
        .replace("[/red]", "\x1b[0m")
        .replace("[/yellow]", "\x1b[0m")
        .replace("[/blue]", "\x1b[0m")
}
